File: python-example/elastic_ingest.py
"""Level 1: pipelined first ingest. SDK batching is automatic; source retention remains your responsibility."""

import os
import time
import uuid
from typing import Optional

os.environ.setdefault("SS_LOG_LEVEL", "warn")
from snowflake.ingest import streaming

DATABASE = os.environ.get("SNOWFLAKE_DATABASE", "MY_DATABASE")
SCHEMA = os.environ.get("SNOWFLAKE_SCHEMA", "MY_SCHEMA")
TABLE = os.environ.get("SNOWFLAKE_TABLE", "MY_TABLE")
PROFILE = os.environ.get("SNOWFLAKE_PROFILE", "profile.json")
RUN_ID = os.environ.get("SNOWFLAKE_RUN_ID", str(uuid.uuid4()))
Row = dict[str, object]

def connection_properties() -> Optional[dict[str, str]]:
    pat = os.environ.get("SNOWFLAKE_PAT")
    if not pat:
        return None
    if not os.environ.get("SNOWFLAKE_ACCOUNT") or not os.environ.get("SNOWFLAKE_URL"):
        raise ValueError("PAT authentication requires SNOWFLAKE_ACCOUNT and SNOWFLAKE_URL")

    properties = {
        "authorization_type": "PAT",
        "personal_access_token": pat,
        "account": os.environ["SNOWFLAKE_ACCOUNT"],
        "url": os.environ["SNOWFLAKE_URL"],
    }
    if os.environ.get("SNOWFLAKE_ROLE"):
        properties["role"] = os.environ["SNOWFLAKE_ROLE"]
    return properties


def create_client() -> streaming.StreamingIngestClient:
    properties = connection_properties()
    return streaming.StreamingIngestClient.from_table(
        client_name=f"ingest-{uuid.uuid4()}",
        db_name=DATABASE,
        schema_name=SCHEMA,
        table_name=TABLE,
        profile_json=None if properties else PROFILE,
        properties=properties,
    )


def sample_row(event_id: int) -> Row:
    return {
        "EVENT_ID": event_id,
        "C1": event_id,
        "C2": f"{RUN_ID}-{event_id}",
    }


def main():
    """Pipeline ten rows, observe every acknowledgement, then close."""
    client = create_client()
    complete = False
    try:
        channel = client.get_elastic_channel()
        # Submit first: waiting after each append would serialize ingestion.
        pending = [channel.append_row_with_wait(sample_row(event_id), None)
                   for event_id in range(10)]
        for acknowledgement in pending:
            acknowledgement.result()
        complete = True
        print(f"Durably acknowledged 10 rows; run={RUN_ID}. Check materialization separately.")
    finally:
        client.close(wait_for_flush=complete, timeout_seconds=30)


if __name__ == "__main__":
    main()

File: nodejs-example/elastic_ingest.js
"use strict";
// Level 1: pipelined first ingest. Sample rows are regenerable; SDK memory is not durable source storage.
process.env.SS_LOG_LEVEL ??= "warn";
const { randomUUID } = require("node:crypto");
const streaming = require("snowpipe-streaming");
const RUN_ID = process.env.SNOWFLAKE_RUN_ID || randomUUID();

function connectionProperties() {
  const pat = process.env.SNOWFLAKE_PAT;
  if (!pat) return null;
  if (!process.env.SNOWFLAKE_ACCOUNT || !process.env.SNOWFLAKE_URL) {
    throw new Error("PAT authentication requires SNOWFLAKE_ACCOUNT and SNOWFLAKE_URL");
  }

  const properties = {
    authorization_type: "PAT",
    personal_access_token: pat,
    account: process.env.SNOWFLAKE_ACCOUNT,
    url: process.env.SNOWFLAKE_URL,
  };
  if (process.env.SNOWFLAKE_ROLE) {
    properties.role = process.env.SNOWFLAKE_ROLE;
  }
  return properties;
}

async function createClient() {
  const properties = connectionProperties();
  return streaming.createTableClient({
    clientName: `ingest-${randomUUID()}`,
    dbName: process.env.SNOWFLAKE_DATABASE || "MY_DATABASE",
    schemaName: process.env.SNOWFLAKE_SCHEMA || "MY_SCHEMA",
    tableName: process.env.SNOWFLAKE_TABLE || "MY_TABLE",
    ...(properties ? { properties } : { profilePath: process.env.SNOWFLAKE_PROFILE || "profile.json" }),
  });
}

function sampleRow(eventId) {
  return {
    EVENT_ID: eventId,
    C1: eventId,
    C2: `${RUN_ID}-${eventId}`,
  };
}

// Pipeline first, then confirm; do not wait after each row.
async function main(clientFactory = createClient) {
  const client = await clientFactory();
  const pending = [];
  let complete = false;
  try {
    const channel = await client.getElasticChannel();
    for (let eventId = 0; eventId < 10; eventId++) {
      pending.push(channel.appendRowWithWait(sampleRow(eventId), null)
        .then(() => null, (error) => error));
    }
    const errors = await Promise.all(pending);
    if (errors.some(Boolean)) throw errors.find(Boolean);
    complete = true;
    console.log(`Durably acknowledged 10 rows; run=${RUN_ID}. Check materialization separately.`);
  } finally {
    await client.close({ waitForFlush: complete, timeoutMs: 30_000 });
  }
}

if (require.main === module) {
  const keepAlive = setInterval(() => {}, 1000);
  main().catch((error) => { console.error(error); process.exitCode = 1; })
    .finally(() => clearInterval(keepAlive));
}
module.exports = { main, createClient, sampleRow };
